Kafka SASL/TLS failures during ProduceSync masked as timeout error after franz-go 1.20.7 upgrade

After upgrading franz-go from v1.18.1 to v1.20.7, SASL authentication failures and TLS errors during `ProduceSync` are no longer returned directly to the client. Instead, franz-go internally retries until the record deadline is exceeded, at which point `kgo.ErrRecordTimeout` is returned. This causes the following error message to be shown to the user:

`records have timed out before they were able to be produced`

when previously the same error would return

`SASL Authentication failed`

I am wondering how do I achieve the previous behavior of surfacing the TLS/SASL errors to the user?

I can see that the error message I want is still being logged, but is no longer captured in the produce result.
```
[kafka] [INFO] immediate metadata update triggered; why: forced load because we are producing to a topic for the first time
[kafka] [INFO] producing to a new topic for the first time, fetching metadata to learn its partitions; topic: s
[kafka] [ERROR] unable to initialize sasl; broker: seed_0, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] new topic metadata wait failed, retrying wait; topic: s, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] metadata update triggered; why: re-updating metadata due to err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [ERROR] unable to initialize sasl; broker: seed_0, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] new topic metadata wait failed, retrying wait; topic: s, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] metadata update triggered; why: re-updating metadata due to err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [ERROR] unable to initialize sasl; broker: seed_0, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] new topic metadata wait failed, retrying wait; topic: s, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] metadata update triggered; why: re-updating metadata due to err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [ERROR] unable to initialize sasl; broker: seed_0, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] new topic metadata wait failed, retrying wait; topic: s, err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] metadata update triggered; why: re-updating metadata due to err: SASL_AUTHENTICATION_FAILED: SASL Authentication failed.: Authentication failed: Invalid username or password
[kafka] [INFO] new topic metadata wait failed, done retrying, failing all records; topic: s, err: records have timed out before they were able to be produced
WARNING: Kafka request timed out. Records have timed out before they were able to be produced <- this is what the user sees
```

This is how I am currently calling `ProduceSync`
```
if err := cl.ProduceSync(ctx, kgo.StringRecord(testMessage)).FirstErr(); err != nil {
		if err == kgo.ErrRecordTimeout {
			err = fmt.Errorf("Kafka request timed out. last err: %w", err)
		}
		return err
}
```
and these are my client opts
```
opts := []kgo.Opt{
		kgo.SeedBrokers([]string{u.Host}...),
		kgo.DefaultProduceTopic(topic),
		kgo.DisableIdempotentWrite(),
		kgo.RecordRetries(5),
		kgo.RecordDeliveryTimeout(20 * time.Second),
		kgo.RetryBackoffFn(backoffFn),
	}
```

--- a/pkg/kgo/producer.go
+++ b/pkg/kgo/producer.go
@@ -1121,6 +1121,7 @@ func (cl *Client) waitUnknownTopic(
 		tries        int
 		unknownTries int64
 		err          error
+		lastRetryErr error
 		after        <-chan time.Time
 	)
 
@@ -1146,7 +1147,11 @@ func (cl *Client) waitUnknownTopic(
 		case <-cl.ctx.Done():
 			err = ErrClientClosed
 		case <-after:
-			err = ErrRecordTimeout
+			if lastRetryErr != nil {
+				err = fmt.Errorf("%w, last err: %w", ErrRecordTimeout, lastRetryErr)
+			} else {
+				err = ErrRecordTimeout
+			}
 		case err = <-unknown.fatal:
 		case retryableErr, ok := <-unknown.wait:
 			if !ok {
@@ -1154,6 +1159,7 @@ func (cl *Client) waitUnknownTopic(
 				return // metadata was successful!
 			}
 			cl.cfg.logger.Log(LogLevelInfo, "new topic metadata wait failed, retrying wait", "topic", topic, "err", retryableErr)
+			lastRetryErr = retryableErr
 			tries++
 			if int64(tries) > cl.cfg.recordRetries {
 				err = fmt.Errorf("no partitions available after attempting to refresh metadata %d times, last err: %w", tries, retryableErr)

